server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -290,30 +305,40 @@ protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext stateme
                         relationContext.parentSources(),
                         coordinatorTxnCtx.sessionSettings().searchPath().currentSchema()),
                     new SubqueryAnalyzer(this, statementContext));
-                Expression expr;
-                if (joinCriteria instanceof JoinOn) {
-                    expr = ((JoinOn) joinCriteria).getExpression();
-                } else {
-                    expr = JoinUsing.toExpression(
-                        leftRel.relationName().toQualifiedName(),
-                        rightRel.relationName().toQualifiedName(),
-                        ((JoinUsing) joinCriteria).getColumns());
-                }
-                try {
-                    joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
-                } catch (RelationUnknown e) {
-                    throw new RelationValidationException(e.getTableIdents(),
-                        String.format(Locale.ENGLISH,
-                        "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
-                }
-            } else {
-                throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "join criteria %s not supported",
-                    joinCriteria.getClass().getSimpleName()));
+                var expressionAnalysisContext = statementContext.currentRelationContext().expressionAnalysisContext();
+                joinCondition = expressionAnalyzer.convert(expression, expressionAnalysisContext);
+            } catch (RelationUnknown e) {
+                throw new RelationValidationException(
+                    e.getTableIdents(),
+                    String.format(Locale.ENGLISH, "missing FROM-clause entry for relation '%s'", e.getTableIdents())
+                );
             }
         }
+        var joinType = JoinType.values()[node.getType().ordinal()];
+        JoinRelation joinRelation = new JoinRelation(leftRel, rightRel, leftRel.outputs(), joinType, joinCondition);
```

Comment:
Does JoinRelation represent a join of left and right? If so, have 2 questions)

1. Why we use `leftRel.outputs()`, and don't take `right.outputs ` as well?

2. Is JoinRelation.outputs used at all? If not, maybe can throw an UnsupportedException there and drop outputs field/ctor parameter?

Reply:
Outputs of Joins are depending on the join, however you are right, only the outputs from the left relation is not correct. So far they are not used yet, but they are part of the `AnalyzedRelation` interface.